test/test_ensemble_builder/test_ensemble_builder.py

```diff
@@ -0,0 +1,554 @@
+      of the dummy runs.
+    """
+    runs = [make_run(loss=10) for _ in range(10)]
+    dummies = [make_run(dummy=True, loss=0) for _ in range(2)]
+
+    candidates, discarded = builder.candidate_selection(
+        runs,
+        dummies,
+        better_than_dummy=True,
+    )
+
+    assert len(candidates) == 2
+    assert all(run.is_dummy() for run in candidates)
+
+
+@parametrize("nbest", [0, 1, 5, 1000])
+def test_candidates_nbest_int(
+    builder: EnsembleBuilder,
+    make_run: Callable[..., Run],
+    nbest: int,
+) -> None:
+    """
+    Expects
+    -------
+    * Should only select the nbest candidates
+    * They should be ordered by loss
+    """
+    n = 10
+    expected = int(bound(nbest, bounds=(1, n)))
+
+    dummy = make_run(dummy=True)
+    runs = [make_run(loss=i) for i in range(n)]
+    candidates, discarded = builder.candidate_selection(runs, dummy, nbest=nbest)
+
+    assert len(candidates) == expected
+
+    if len(candidates) > 1:
+        assert all(a.loss <= b.loss for a, b in pairs(candidates))
+
+    if any(discarded):
+        worst_candidate = candidates[-1]
+        assert all(worst_candidate.loss <= d.loss for d in discarded)
+
+
+@parametrize("nbest", [0.0, 0.25, 0.5, 1.0])
+def test_candidates_nbest_float(
+    builder: EnsembleBuilder,
+    make_run: Callable[..., Run],
+    nbest: float,
+) -> None:
+    """
+    Expects
+    -------
+    * Should select nbest percentage of candidates
+    * They should be ordered by loss
+    """
+    n = 10
+    expected = int(bound(nbest * n, bounds=(1, n)))
+
+    dummy = make_run(dummy=True, loss=0)
+    runs = [make_run(id=i, loss=i) for i in range(2, n + 2)]
+    candidates, discarded = builder.candidate_selection(runs, dummy, nbest=nbest)
+
+    assert len(candidates) == expected
+
+    if len(candidates) > 1:
+        assert all(a.loss <= b.loss for a, b in pairs(candidates))
+
+    if any(discarded):
+        worst_candidate = candidates[-1]
+        assert all(worst_candidate.loss <= d.loss for d in discarded)
+
+
+@parametrize("threshold", [0.0, 0.25, 0.5, 1.0])
+def test_candidates_performance_range_threshold(
+    builder: EnsembleBuilder,
+    make_run: Callable[..., Run],
+    threshold: float,
+) -> None:
+    """
+    Expects
+    -------
+    * Should select runs that are `threshold` between the dummy loss and the best loss
+      This value is captured in `boundary`.
+    """
+    worst_loss = 100
+    best_loss = 0
+    dummy_loss = 50
+
+    boundary = threshold * best_loss + (1 - threshold) * dummy_loss
+
+    dummy = make_run(dummy=True, loss=dummy_loss)
+    runs = [make_run(loss=loss) for loss in np.linspace(best_loss, worst_loss, 101)]
+
+    candidates, discarded = builder.candidate_selection(
+        runs,
+        dummy,
+        performance_range_threshold=threshold,
+    )
+
+    # When no run is better than threshold, we just get 1 candidate,
+    # Make sure it's the best
+    if len(candidates) == 1:
```

Comment:
There's no check for zero candidates here.

Reply:
Good spot, I threw in `assert len(candidates) > 0`